### PR TITLE
Alamofire 5 Core Rewrite!

### LINKED ISSUE #2226: Asynchronous Request Creation
Perhaps the most inelegant bit of Alamofire's current API is the handling of multipart form uploads. Due to the asynchronous nature of the request creation, it doesn't support chaining like every other `Request` API, requiring users to get the actual upload request in the completion for the request creation. Alamofire should move to an asynchronous `Request` creation model for all request types. Not only could this greatly simplify multipart uploads, but would allow expensive operations in other requests as well, an issue we've had reported before. Most likely this will entail something like a `Future` API for `Request`, or at least an internal state machine so that method chaining can still work.

### PR DESCRIPTION
**Pull request by jshier**
### Goals :soccer:
This PR is a complete rewrite of core Alamofire types and behaviors. Fixes #2226, #2235, #2225, #2300, among others.

### Needed :white_check_mark:
- [x] Finalize which `EventMonitor` types will be provided, reimplement notifications as an `EventMonitor`, add `EventMonitor` tests.
- [x] Reimplement `description` and `debugDescription`
- [x] Reimplement `Response` mapping methods.
- [ ] Update all inline documentation.

### Implementation Details :construction:
tl;dr: `SessionManager`, `SessionDelegate`, and the `Request` types have been completely rewritten and most other areas of the library touched in some way. Oh, and iOS 10 is now the minimum (unless the `URLSessionTaskMetrics` linker bug on watchOS forces us to backtrack on metrics support).

#### Summary
This PR rewrites `SessionManager`, `SessionDelegate`, and the `Request` types for many reasons, including:

- Library evolution had lead to functionality and areas of responsibility being spread amongst many types within Alamofire. This made it increasingly difficult to add features, diagnose and fix bugs, and optimize performance. Most notably, the original subdelegate architecture has been completely removed, as it was a huge source of complexity and one of the biggest things preventing a Linux port. The `TaskDelegate` functionality has been reimplemented in the `SessionDelegate` or `Request`s themselves.
- Queue usage was distinctly suboptimal, with separate queues for the `URLSessionDelegate` and each `Request`, as well as global queue usage for things like multipart form requests and retry, easily leading to scenarios where users were able to negatively impact system performance without realizing it and without being able to do anything about it.
- Swift has evolved substantially over the lifetime of Alamofire and Alamofire hasn't kept pace. This is especially true in the area of error handling, where we using own custom constructs where standard `try/catch` would suffice. This rewrite updates those areas to better follow Swift convention and the advancements made over the years.
- Alamofire's usage of the `URLSessionDelegate` methods was preventing implementation of a few features and was burdened by trying to cover all facets of `URLSession`. This rewrite both prunes features that weren't tested and weren't likely being used (streaming) and reimplements security challenge handling by changing which delegate methods are implemented.

#### `SessionManager`
`SessionManager` has been reimplemented in order to simplify it's two areas of responsibility:

- Initializing dependent parts of Alamofire.
- Creating `Request`s.

...

### Testing Details :mag:
Existing tests have been updated but are largely untouched.

### CODE REVIEW

File: Source/Alamofire.swift [hunk 1]
```diff
@@ -116,303 +116,309 @@ extension URLRequest {
...
-    usingThreshold encodingMemoryThreshold: UInt64 = MultipartUpload.multipartFormDataEncodingMemoryThreshold,
-    with urlRequest: URLRequestConvertible) -> UploadRequest
-{
-    return SessionManager.default.upload(
-        multipartFormData: multipartFormData,
-        usingThreshold: encodingMemoryThreshold,
-        with: urlRequest
-    )
+/// Global namespace containing API for the `default` `SessionManager` instance.
+public enum Alamofire {
```
Comment (GoCac) [inline Source/Alamofire.swift]: Hi, @jshier ,  there is a question - why do you choice `enum` rather than `struct` here？I wonder to know the difference between them. Thank you.

Comment (jshier) [reply Source/Alamofire.swift]: By convention, since Swift doesn’t have namespaces, enums are used instead, as they can’t be accidentally instantiated.

Comment (GoCac) [reply Source/Alamofire.swift]: got it. Thanks for your reply~👍

**Comment by freak4pc** [comment]
Hey! I'm aware this is an old PR, but any whether or not this is documented somehow, migration-wise? I'm trying to figure out some of the latest changes to make sure everything is still ticking for us.
